src/Server/Sdk/Bindings/WebSocketChannelListener.cs

```diff
@@ -0,0 +1,69 @@
+﻿using System.Diagnostics;
+using System.Net.WebSockets;
+using System.Reactive.Linq;
+using System.Reactive.Subjects;
+
+namespace System.Net.Mqtt.Sdk.Bindings
+{
+	/// <summary>
+	/// Represents a listener for incoming channel connections on top of Web Sockets
+	/// The connections are performed by MQTT Clients through <see cref="IMqttChannel{T}" /> 
+	/// and the listener accepts and establishes the connection on the Server side
+	/// </summary>
+	public class WebSocketChannelListener : IMqttChannelListener
+	{
+		static readonly ITracer tracer = Tracer.Get<WebSocketChannelListener> ();
+		static readonly Subject<WebSocket> listener = new Subject<WebSocket> ();
+
+		readonly MqttConfiguration configuration;
+		bool disposed;
+
+		internal WebSocketChannelListener (MqttConfiguration configuration)
+		{
+			this.configuration = configuration;
+		}
+
+		/// <summary>
+		/// Accepts an incoming client request as a Web Socket
+		/// </summary>
+		/// <param name="webSocket">The Web Socket instance to accept</param>
+		public static void AcceptWebSocketClient (WebSocket webSocket) => listener.OnNext (webSocket);
+
+#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member
+		public IObservable<IMqttChannel<byte[]>> GetChannelStream ()
```

Comment:
Any particular reason *not* to document this member? If the intention is not to because it's part of `IMqttChannelListener`, then maybe it should be implemented explicitly?

Reply:
In general, the listeners are not exposed to the public API, because it's something more low level. You just interact with the Bindings when creating the Client or Server. Then the bindings interacts with Channel Factories and Listeners.

The thing is that for the case of Web Sockets, there is no direct method to accept a web socket like the TcpListener does, so what I did is to expose this class without a public ctor and add a static method to link the WebSocket instance with the listener.
So. it's like a manual way to link the current WebSocket to the infrastructure.
Depending on what you are using to get the WebSockets (ASP.net, etc), the way to obtain it is different, so that's why I don't take a dependency on the way to obtain the instance but I just let the consumer to assign it.

For that need of exposing the listener I just need to implement the other interace methods as public. But I agree that implementing them explicitly would be a better choice.

Reply:
Hm, I totally missed that, since we don't have a unit test that showcases the typical usage :(. Is there a simple-ish way we can spin up a websockets server in a test and connect to it from the test, just like we do for other end to end tests?